onnxruntime/contrib_ops/cuda/custom_layer_norm.cu

```diff
@@ -0,0 +1,307 @@
+
+  for (int stride = 1; stride < GPU_WARP_SIZE; stride *= 2) {
+    variance += WARP_SHFL_DOWN(variance, stride);
+  }
+
+  if (g.thread_rank() == 0) shr[gid] = variance;
+
+  b.sync();
+
+  if (g.thread_rank() < (iteration_stride >> GPU_WARP_SIZE_LOG2)) variance = shr[g.thread_rank()];
+
+  b.sync();
+
+  for (int stride = 1; stride < (iteration_stride >> GPU_WARP_SIZE_LOG2); stride *= 2) {
+    variance += WARP_SHFL_DOWN(variance, stride);
+  }
+  variance = WARP_SHFL(variance, 0);
+  variance /= (row_stride * 2);
+  variance += epsilon;
+  const float inv_variance = rsqrt(variance);
+
+  __half2 mean_h = __float2half2_rn(mean);
+  __half2 variance_h = __float2half2_rn(variance);
+  const __half2* gamma_cast = reinterpret_cast<const __half2*>(gamma);
+  const __half2* beta_cast = reinterpret_cast<const __half2*>(beta);
+
+  if (g.thread_rank() == 0) {
+    invvars[row] = inv_variance;
+    means[row] = mean;
+  }
+
+  for (int i = 0; i < iterations; i++) {
+    vals_arr[i] = __float22half2_rn(vals_f[i]);
+    vals_arr[i] = (vals_arr[i] - mean_h) * h2rsqrt(variance_h);
+    vals_arr[i] = vals_arr[i] * gamma_cast[i * iteration_stride + id] +
+                  beta_cast[i * iteration_stride + id];
+    vals_cast[row * row_stride + i * iteration_stride + id] = vals_arr[i];
+  }
+#endif
+}
+
+template <typename T, typename U>
+void launch_custom_layer_norm(
+    T* vals,
+    const T* residual,
+    const T* gamma,
+    const T* beta,
+    U epsilon,
+    int64_t n1,
+    int64_t n2,
+    U* invvars,
+    U* means);
+
+template <>
+void launch_custom_layer_norm<float, float>(
+    float* vals,
+    const float* residual,
+    const float* gamma,
+    const float* beta,
+    float epsilon,
+    int64_t n1,
+    int64_t n2,
+    float* invvars,
+    float* means) {
+  constexpr int threads = 256;
+
+  dim3 grid_dim(n1);
+
+  dim3 block_dim(threads);
+
+  // There are some limitations to call below functions, now just enumerate the situations.
+  if (n2 == 768)
+    custom_layer_norm<768, 3><<<grid_dim, block_dim, 0, 0>>>(
+        vals, residual, gamma, beta, epsilon, invvars, means);
+  else if (n2 == 512)
+    custom_layer_norm<512, 2><<<grid_dim, block_dim, 0, 0>>>(
+        vals, residual, gamma, beta, epsilon, invvars, means);
+  else if (n2 == 1024)
+    custom_layer_norm<1024, 4><<<grid_dim, block_dim, 0, 0>>>(
+        vals, residual, gamma, beta, epsilon, invvars, means);
+  else if (n2 == 1536)
+    custom_layer_norm<1536, 6><<<grid_dim, block_dim, 0, 0>>>(
+        vals, residual, gamma, beta, epsilon, invvars, means);
+  else if (n2 == 2048)
+    custom_layer_norm<2048, 8><<<grid_dim, block_dim, 0, 0>>>(
+        vals, residual, gamma, beta, epsilon, invvars, means);
+  else if (n2 == 2560)
+    custom_layer_norm<2560, 10><<<grid_dim, block_dim, 0, 0>>>(
+        vals, residual, gamma, beta, epsilon, invvars, means);
+  else {
+    //shouldn't reach here
+    ORT_THROW("Unsupported hidden_dim for custom LayerNorm implementation.");
+  }
```

Comment:
Can the code above be changed like below? Then it is not necessary to specialized for half.
if (n2 == 768)
    custom_layer_norm<**768 * sizeof(T) / 4**, 3><<<grid_dim, block_dim, 0, 0>>>(
        vals, residual, gamma, beta, epsilon, invvars, means);
  else if (n2 == 512)
    custom_layer_norm<**512* sizeof(T) / 4**, 2><<<grid_dim, block_dim, 0, 0>>>(
        vals, residual, gamma, beta, epsilon, invvars, means);
  else if (n2 == 1024)
    ....
  else {
    //shouldn't reach here
    ORT_THROW("Unsupported hidden_dim for custom LayerNorm implementation.");
  }